### ISSUE TITLE
sulogin is not being installed

### ISSUE THREAD

**Issue by firasuke** [shadow-maint/shadow#1082]
Hey there,

Thank for your time and effort.

I noticed that shadow's `sulogin` is being built under `src/`, but it is not being installed when running `make install`. Having checked `src/Makefile` it seems that it is under `noinst_PROGRAMS`.

Is this intended behavior? Does this mean that util-linux's `sulogin` is the version to use?

Thanks in advance!

**Comment by alejandro-colomar** [comment]
> Is this intended behavior?

Having an explicit line in the Makefile.am, I'd say yes, it was intended.  But I don't know _why_ it was intended.  The git-blame(1) says it's a very old line.

> Does this mean that util-linux's `sulogin` is the version to use?

Probably.  You could check what distros are doing.

**Comment by firasuke** [comment]
> > Does this mean that util-linux's `sulogin` is the version to use?
> 
> Probably. You could check what distros are doing.

`util-linux`'s version of `sulogin` is indeed what is being used by almost all distributions. I guess `shadow`'s `sulogin` could be dropped then?

**Comment by alejandro-colomar** [comment]
> > > Does this mean that util-linux's `sulogin` is the version to use?
> > 
> > 
> > Probably. You could check what distros are doing.
> 
> `util-linux`'s version of `sulogin` is indeed what is being used by almost all distributions.

re: almost: Did you find any one that doesn't?

> I guess `shadow`'s `sulogin` could be dropped then?

It depends on your response to the above question.  If _none_ at all are using it, maybe it would make sense to remove it, since I wouldn't trust a program that nobody has been using at all for years.  If there's any user, I guess we should keep it.

**Comment by firasuke** [comment]
> re: almost: Did you find any one that doesn't?

After a basic search on publicly available code on GitHub, I found some [slackware recipes](https://github.com/nobodino/slackware-from-scratch/blob/master/musl/a/shadow/shadow.SlackBuild#L217) and some [Chimera Linux recipes](https://github.com/chimera-linux/cports/blob/master/main/shadow/template.py#L59) that are explicitly installing `shadow`'s version of `sulogin`.

**Comment by alejandro-colomar** [comment]
If there are users, I'd prefer keeping it.

**Comment by firasuke** [comment]
> > re: almost: Did you find any one that doesn't?
> 
> Yes, after a basic search on publicly available code on GitHub, I found some [slackware recipes](https://github.com/nobodino/slackware-from-scratch/blob/master/musl/a/shadow/shadow.SlackBuild#L217) and some [Chimera Linux recipes](https://github.com/chimera-linux/cports/blob/master/main/shadow/template.py#L59) that are explicitly installing `shadow`'s version of `sulogin`.


> If there are users, I'd prefer keeping it.

@q66 @nobodino would it be possible for you to switch to `util-linux`'s version of `sulogin` so that `shadow`'s relatively "limited" version can be deprecated and removed peacefully? It is already not intended for installation, and I am willing to cooperate hopefully with relevant PRs to both your `shadow` and `util-linux` recipes.

**Comment by alejandro-colomar** [comment]
> > > re: almost: Did you find any one that doesn't?
> > 
> > 
> > Yes, after a basic search on publicly available code on GitHub, I found some [slackware recipes](https://github.com/nobodino/slackware-from-scratch/blob/master/musl/a/shadow/shadow.SlackBuild#L217) and some [Chimera Linux recipes](https://github.com/chimera-linux/cports/blob/master/main/shadow/template.py#L59) that are explicitly installing `shadow`'s version of `sulogin`.
> 
> > If there are users, I'd prefer keeping it.
> 
> [@q66](https://github.com/q66) [@nobodino](https://github.com/nobodino) would it be possible for you to switch to `util-linux`'s version of `sulogin` so that `shadow`'s relatively "limited" version can be deprecated and removed peacefully? It is already not intended for installation, and I am willing to cooperate hopefully with relevant PRs to both your `shadow` and `util-linux` recipes.

Could you clarify what you mean by limited?

If you mean that it has less features, that might in itself be a feature.  Being simpler means it might be safer.

I'm not a fan of deprecating and removing an implementation that works fine.  Diversity can be good evolutively.

Also, switching to a different implementation might come with regressions.  <https://github.com/shadow-maint/shadow/issues/999#issuecomment-2131309982>

**Comment by firasuke** [comment]
> Could you clarify what you mean by limited?
> 
> If you mean that it has less features, that might in itself be a feature. Being simpler means it might be safer.
> 
> I'm not a fan of deprecating and removing an implementation that works fine. Diversity can be good evolutively.
> 
> Also, switching to a different implementation might come with regressions. [#999 (comment)](https://github.com/shadow-maint/shadow/issues/999#issuecomment-2131309982)

Oh I did not mean it in a bad way. I meant that `util-linux`'s is the more widespread version with support for option flags and what not. I am a fan of the smaller "more limited" version (I even prefer the versions from `toybox`/`busybox`) but it makes no sense to have multiple providers of the same utility.

I have opened a [PR](https://github.com/chimera-linux/cports/pull/5066) at Chimera's cports regarding this issue (also it appears that they [intended to use the version from `util-linux` already](https://github.com/chimera-linux/dinit/blob/master/doc/linux/services/recovery#L2)?).

**Comment by q66** [comment]
chimera does not explicitly depend on anything from util-linux except for mount utils right now and i'd prefer to keep it that way

the shadow sulogin works fine and doesn't need any more features considering it's not something ever invoked by a user and serves a single purpose

**Comment by alejandro-colomar** [comment]
> > Could you clarify what you mean by limited?
> > If you mean that it has less features, that might in itself be a feature. Being simpler means it might be safer.
> > I'm not a fan of deprecating and removing an implementation that works fine. Diversity can be good evolutively.
> > Also, switching to a different implementation might come with regressions. [#999 (comment)](https://github.com/shadow-maint/shadow/issues/999#issuecomment-2131309982)
> 
> Oh I did not mean it in a bad way. I meant that `util-linux`'s is the more widespread version with support for option flags and what not. I am a fan of the smaller "more limited" version (I even prefer the versions from `toybox`/`busybox`) but it makes no sense to have multiple providers of the same utility.

I'm not sure I understand the last sentence.  busybox/toybox/coreutils are different providers of the same utilities, yet you seem to like the existence of all of them.

As a Devuan user, I'm rather opposed to unification of providers of tools (init(8)).

And even if for libc where I prefer an abundant library, and use glibc myself, I love that musl exists.

I'm happy to remove tools that are not anymore (logoutd(8)) or are even actively harmful (expiry(1), chage(1)).  But I'm inclined to not remove tools just because they're provided by other projects.

**Comment by firasuke** [comment]
> As a Devuan user, I'm rather opposed to unification of providers of tools (init(8)).

Again, I am with you on this one, but it does not make sense to deprecate utilities like `id` and `groups` citing the existence of replacements from `coreutils`, and then not do the same for `sulogin`

> The shadow implementations of id(1) and groups(1) are deprecated in favor of the GNU coreutils and binutils versions. They will be removed in 4.17.0.

I am just pointing out the discrepancy here..

> I'm happy to remove tools that are not used, or tools that are not used anymore (logoutd(8)) or even actively harmful (expiry(1), chage(1)).

Sure, I am also happy to see that new releases of shadow are indeed smaller, but when `id` and `groups` were removed they had more users, unlike `sulogin`.

Thank you again for your time and effort.

**Comment by alejandro-colomar** [comment]
> > As a Devuan user, I'm rather opposed to unification of providers of tools (init(8)).
> 
> Again, I am with you on this one, but it does not make sense to deprecate utilities like `id` and `groups` citing the existence of replacements from `coreutils`, and then not do the same for `sulogin`
> 
> > The shadow implementations of id(1) and groups(1) are deprecated in favor of the GNU coreutils and binutils versions. They will be removed in 4.17.0.
> 
> I am just pointing out the discrepancy here..
> 
> > I'm happy to remove tools that are not used, or tools that are not used anymore (logoutd(8)) or even actively harmful (expiry(1), chage(1)).
> 
> Sure, I am also happy to see that new releases of shadow are indeed smaller, but when `id` and `groups` were removed they had more users, unlike `sulogin`.

Did they really have more users?  Nobody reported that, from what I can read in the PR and issue.
-  <https://github.com/shadow-maint/shadow/issues/1005>
-  <https://github.com/shadow-maint/shadow/pull/1008>

If we had users, we would have considered keeping them.

Maybe they were unaware, though.  @q66 did that removal cause any issues for you?

Another thing to keep in mind is that groups(1) and id(1) are just for printing stuff to stdout, so it's not too worrying if you change the provider to coreutils.  On the other hand, sulogin(8) does much more than that, and regressions in the program can be quite dangerous.

> 
> Thank you again for your time and effort.

Thank you!

**Comment by q66** [comment]
> Maybe they were unaware, though. [@q66](https://github.com/q66) did that removal cause any issues for you?

no, we use the freebsd versions of those in our core userland package

**Comment by firasuke** [comment]
I think a better idea would be to provide a configuration flag `--without-sulogin` that prevents `sulogin` from being built.

The current status quo, the topic of this issue, is not considered an expected behavior; `sulogin` is being built implicitly regardless of the use case.

**Comment by alejandro-colomar** [comment]
> I think a better idea would be to provide a configuration flag `--without-sulogin` that prevents `sulogin` from being built.
> 
> The current status quo, the topic of this issue, is not considered an expected behavior; `sulogin` is being built implicitly regardless of the use case.

See <https://github.com/shadow-maint/shadow/issues/999#issuecomment-2124307068>.
